File: docs/article/source/scala/apps/cargo-tracker/app/cargotracker/handling/domain/model/repositories/HandlingActivityRepository.scala
package cargotracker.handling.domain.model.repositories

import cargotracker.handling.domain.model.aggregates.HandlingActivity

trait HandlingActivityRepository:

  /** 自前で `DB.localTx` を開いて save する (既存呼出側互換)。 */
  def save(activity: HandlingActivity): Unit

  /** IT9 0.1 (ADR 0016 案 A): 外側 TX に参加して save する。
    *
    *   - HandlingOrchestrator が外側で `DB.localTx { implicit session => ... }` を開く場合に使う
    *   - ステップ間失敗時の rollback (Handling + Tracking + Booking 連動) を可能にする
    */
  def saveInTx(activity: HandlingActivity)(implicit session: scalikejdbc.DBSession): Unit

  def findByBookingId(bookingId: String): Seq[HandlingActivity]
  def findAll(): Seq[HandlingActivity]

File: apps/cargo-tracker/src/main/java/com/example/cargotracker/tracking/handling/domain/repository/HandlingActivityRepository.java
package com.example.cargotracker.tracking.handling.domain.repository;

import com.example.cargotracker.tracking.handling.domain.model.CargoBookingId;
import com.example.cargotracker.tracking.handling.domain.model.HandlingActivity;
import java.util.List;

/** 荷役作業の出力ポート。実装はインフラ層に置く（DIP）。 */
public interface HandlingActivityRepository {

    /** 荷役作業を登録する（US15）。 */
    void save(HandlingActivity activity);

    /** 予約ごとの荷役履歴を、作業日時の新しい順で返す。 */
    List<HandlingActivity> findByBookingId(CargoBookingId bookingId);

    /**
     * 荷役履歴を新しい順で返す（荷役作業一覧）。
     *
     * <p><strong>登録した作業を先頭に表示する</strong>ため、既定の並びは
     * 作業日時の新しい順である（{@code ui_design.md}）。
     *
     * @param limit 取得件数の上限
     */
    List<HandlingActivity> findRecent(int limit);
}
